Add Conversation AI to Java SDK

siri-varma opened the pull request:
# PR Overview: Conversation AI SDK Integration

This PR introduces the **Conversation AI SDK** with the **Converse API**. The implementation is structured into the following categories:

- **Core Logic**
- **Models**
- **Unit Tests**

---

## Core Logic

The core logic is implemented in a key class, as detailed below:

| **Class**                  | **Description**                                                                                       | **Details**                                                                                                         |
|----------------------------|-------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------|
| **DaprConversationClient**  | Manages the Conversation API. Handles validations and async calls.                                    | - Calls the Conversation API asynchronously. <br> - Validates inputs and manages job execution. |

---

## API Contracts

The following is the method signature for the **Dapr Converse API**:

### **Dapr Converse API**

This method interacts with the Dapr Converse API.

```java
Mono<DaprConversationResponse> converse(
    String conversationComponentName,
    List<DaprConversationInput> daprConversationInputs,
    String contextId,
    boolean scrubPii,
    double temperature
);
```

| **Parameter**                | **Description** |
|------------------------------|----------------|
| `conversationComponentName`  | Name of the conversation component. |
| `daprConversationInputs`     | List of prompts that are part of the conversation. |
| `contextId`                  | Identifier for an existing chat session (e.g., similar to ChatGPT). |
| `scrubPii`                   | Whether to scrub personally identifiable information (PII) from responses. |
| `temperature`                | Controls response randomness (lower = more predictable, higher = more creative). |
| **Returns**                  | `DaprConversationResponse` |

---

## Models

The SDK follows the **builder pattern** for constructing models, ensuring cleaner and more maintainable object creation.

---

## Tests

The testing strategy includes:

- **Unit Tests**: Validate individual components and business logic.

---

## Issue Reference

We ensure that all PRs are linked to a relevant issue where the problem or feature has been discussed before implementation.

This PR **closes** the following issue: **#1101**

---

## Checklist ✅

Please confirm the following before merging:

- [X] Code compiles correctly
- [X] Created/updated tests
- [X] Extended the documentation

---

siri-varma commented:
@salaboy please let me know what you think. Working on integration and unit tests in the meanwhile

yaron2 commented:
Great contribution, thanks @siri-varma

salaboy commented:
this is great @siri-varma ! I will start reviewing soon!

siri-varma commented:
@salaboy For integration tests, we must spin up a mock for third party LLM endpoint. Do we have similar patterns in the current setup? (where dapr is talking to a third party api and receiving responses)

salaboy commented on sdk-ai/src/main/java/io/dapr/ai/client/DaprConversationClient.java:
```diff
@@ -0,0 +1,198 @@
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class DaprConversationClient implements AutoCloseable, DaprAiClient {
```
@siri-varma I assume that you know that we had a previewClient, why is this introducing a new separated client for the conversation API? I would assume that we want to include this in PreviewClient first. @yaron2 we might need your help here.

salaboy replied:
and sorry for the confusion if the PreviewClient is an old thing that we are not supposed to use anymore. As the workflows now have their own separate client, this might be the way to go.

siri-varma replied:
Just to clarify we are fine with current approach right ?

salaboy commented on sdk-ai/pom.xml:
```diff
@@ -0,0 +1,164 @@
+    <artifactId>dapr-sdk-ai</artifactId>
+    <packaging>jar</packaging>
+    <version>1.15.0-SNAPSHOT</version>
+    <name>dapr-sdk-ai</name>
```
@siri-varma should we call it conversation? instead of `ai`? as `ai` is too generic

siri-varma replied:
Changed to conversation

salaboy commented on sdk-ai/pom.xml:
```diff
@@ -0,0 +1,164 @@
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
```
please include the nexus plugin as shown in here: https://github.com/dapr/java-sdk/blob/master/dapr-spring/pom.xml#L95

siri-varma replied:
added the nexus plugin

salaboy commented:
@siri-varma this is looking good, I left some comments that basically highlight my doubts about some JDK general topics. But this is looking really good, are you planning to add integration tests inside the `sdk-tests` too? We have some Spring Boot tests using Testcontainers, I was thinking that you might want to test the APIs against a local model running in a docker container to see that everything works as expected.

siri-varma commented:
@salaboy I have tested the code by running it locally and using my OpenAI API Key. 

Regarding the integration test, I think we will need an LLM model for daprd to communicate with, which unfortunately are huge in size. That is why I did not add the integration tests.

salaboy commented:
Yeah check this https://java.testcontainers.org/modules/ollama/

siri-varma commented:
@salaboy thank you for providing the testcontainer urls.

Like how we have url overrides for Scheduler, I was looking one for conversation too because we will have to point dapr to the local llm model. But could not find any

I tried looking in the below places and could not find anything. https://github.com/dapr/dapr/blob/2e7c61e933099b9d40fcdefbd57cc3e81069915a/pkg/injector/annotations/annotations.go#L73
https://github.com/dapr/dapr/blob/2e7c61e933099b9d40fcdefbd57cc3e81069915a/pkg/injector/consts/consts.go#L51
https://docs.dapr.io/reference/arguments-annotations-overview/

@salaboy  or @cicoyle Would you folks know about this ?

artur-ciocanu commented:
@siri-varma thanks a lot for your contribution. I have reviewed your PR and I have a few comments similar to #1255:
- Let's make sure the Conversation API is part of the `DaprPreviewClient`
- Let's make sure that we use the `Alpha1` suffix for the Conversation API
- Let's make sure we keep everything under `sdk` module and we don't create a separate module for Conversation API
- For the Conversation API `converse` let's make sure we add private methods to validate all the input parameters, before invoking the GRPC API
- For Conversation API `converse` method let's ensure we keep number of parameters to a minimum, having to pass 5 parameters is too much. I think a better option is to have a builder for `ConversationRequest` which should be part of the model classes for `DaprClient`. This way we can have sane defaults, but allow users to override some of the values if desired.

CC: @salaboy @cicoyle

artur-ciocanu reviewed (commented):
@siri-varma thanks a lot for your contribution, I have left a longer comment with some of my thoughts.

Please take a look and let me know what you think.

siri-varma commented:
> @siri-varma thanks a lot for your contribution. I have reviewed your PR and I have a few comments similar to #1255:
> 
> * Let's make sure the Conversation API is part of the `DaprPreviewClient`
> * Let's make sure that we use the `Alpha1` suffix for the Conversation API
> * Let's make sure we keep everything under `sdk` module and we don't create a separate module for Conversation API
> * For the Conversation API `converse` let's make sure we add private methods to validate all the input parameters, before invoking the GRPC API
> * For Conversation API `converse` method let's ensure we keep number of parameters to a minimum, having to pass 5 parameters is too much. I think a better option is to have a builder for `ConversationRequest` which should be part of the model classes for `DaprClient`. This way we can have sane defaults, but allow users to override some of the values if desired.
> 
> CC: @salaboy @cicoyle

@artur-ciocanu Addressed all the comments here as well.
@salaboy Added Integration tests too.

siri-varma commented:
@artur-ciocanu , @salaboy All the builds are passing and addressed comments as well. Would you folks be able to do another round of reviews please ?

cicoyle commented on examples/src/main/java/io/dapr/examples/conversation/DemoConversationAI.java:
```diff
@@ -0,0 +1,35 @@
+package io.dapr.examples.conversation;
```
Please ensure all files have the header 🙏🏻

siri-varma replied:
Done.

cicoyle commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprConversationIT.java:
```diff
@@ -0,0 +1,119 @@
+    @BeforeEach
+    public void setUp(){
+        org.testcontainers.Testcontainers.exposeHostPorts(PORT);
+        // Ensure the subscriptions are registered
```
is this an intentional comment?
```suggestion
```

cicoyle reviewed (changes requested):
Generally this looks great 🌟 - mind adding an example for conversation api to the examples directory in the repo with the mechanical markdown in the README with a sample .java file for users to know how to use the API?

We should also document the high level things that are expected when adding new APIs to the SDK to lower the barrier to entry for future developers and for consistency. Thanks for this PR 🚀 We are nearly there!

cicoyle commented:
Amazing work here @siri-varma 🚀 Thanks for the contribution 🌟

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationInput.java:
```diff
@@ -0,0 +1,84 @@
+  private final String content;
+
+  private ConversationRole role;
+
```
Can we use something more user friendly like `removeSensitiveInformation` instead of `scrubPii`. I think the same should apply to Dapr Runtime, but at least at the SDK level we should be more user friendly.

@cicoyle what are your thoughts on this.

siri-varma replied:
using scrubPii as discussed below

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationOutput.java:
```diff
@@ -0,0 +1,56 @@
+   *
+   * @return parameters.
+   */
+  public Map<String, byte[]> getParameters() {
```
You can use `Map.of()`

siri-varma replied:
Because this was an output did not want users to modify it

siri-varma replied:
Changed to Map.copyOf

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationRequest.java:
```diff
@@ -0,0 +1,119 @@
+ * conversation inputs, context identifier, PII scrubbing, and temperature control.
+ */
+public class ConversationRequest {
+
```
Could we use `name` or `componentName` instead of `llmName`? I have checked the Dapr docs here: https://docs.dapr.io/developing-applications/building-blocks/conversation/howto-conversation-layer/ and as far as I can tell `name` refers to Dapr Component name, not LLM name.

I would suggest opening an issue with Dapr Core to rename this, since `name` is way to generic.

CC: @cicoyle

cicoyle replied:
The request should contain the generic `name` but internally we can use `llmname`, [see how the go-sdk does that here](https://github.com/dapr/go-sdk/blob/a3df75f17b648c0de4642fb67c575f3d34d2dc62/client/conversation.go#L38). We should definitely document what `name` is - as it can be a bit convoluted.

siri-varma replied:
changed to use `name` and added better documentation

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationRequest.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class ConversationRequest {
+
+  private final String llmName;
```
Any reason this is not just `conversationInputs`?

siri-varma replied:
changed to inputs

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationRequest.java:
```diff
@@ -0,0 +1,119 @@
+
+  private final String llmName;
+  private final List<ConversationInput> daprConversationInputs;
+  private String contextId;
```
As I mentioned could we have something more user friendly, scrub PII seems very technical

siri-varma replied:
using scrubPii because checkstyle is complaining about using caputals `II`

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationRequest.java:
```diff
@@ -0,0 +1,119 @@
+    this.temperature = temperature;
+    return this;
+  }
+}
```
Please add a new line

siri-varma replied:
done

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationResponse.java:
```diff
@@ -0,0 +1,56 @@
+public class ConversationResponse {
+
+  private String contextId;
+
```
Same here, any reason we can't use just `conversationOutputs`?

cicoyle replied:
lets use `outputs` here to align with the go-sdk and dotnet-sdk.

siri-varma replied:
using outputs

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationResponse.java:
```diff
@@ -0,0 +1,56 @@
+   *
+   * @return List{@link ConversationOutput}.
+   */
+  public List<ConversationOutput> getConversationOutpus() {
```
Use `List.copyOf()` instead of `Collections.unmodifiableList()`

siri-varma replied:
The intention was to make sure the response cannot be modified so we have used unmodifiedlist.

siri-varma replied:
Changed to List.copyOf

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
 	}
+
+	@Test
+	public void converseShouldThrowIllegalArgumentExceptionWhenComponentNameIsNull() throws Exception {
```
`daprConversationInputs` could we rename it to just `inputs`?

siri-varma replied:
changed to inputs

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+	}
+
+	@Test
+	public void converseShouldThrowIllegalArgumentExceptionWhenConversationComponentIsEmpty() throws Exception {
```
Same here

siri-varma replied:
changed to inputs

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+	}
+
+	@Test
+	public void converseShouldThrowIllegalArgumentExceptionWhenConversationInputIsEmpty() throws Exception {
```
Same here

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+	}
+
+	@Test
+	public void converseShouldThrowIllegalArgumentExceptionWhenConversationInputContentIsNull() throws Exception {
```
Same here

siri-varma replied:
changed to inputs

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+	}
+
+	@Test
+	public void converseShouldThrowIllegalArgumentExceptionWhenConversationInputContentIsEmpty() throws Exception {
```
Same here

siri-varma replied:
changed to inputs

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+			observer.onCompleted();
+			return null;
+		}).when(daprStub).converseAlpha1(any(DaprProtos.ConversationRequest.class), any());
+
```
Same here

siri-varma replied:
done

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+		}).when(daprStub).converseAlpha1(any(DaprProtos.ConversationRequest.class), any());
+
+		List<ConversationInput> daprConversationInputs = new ArrayList<>();
+		daprConversationInputs.add(new ConversationInput("Hello there"));
```
Could it be just `response`

siri-varma replied:
changed to response

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+			observer.onCompleted();
+			return null;
+		}).when(daprStub).converseAlpha1(any(DaprProtos.ConversationRequest.class), any());
+
```
Could we rename it to `daprConversationInput`

siri-varma replied:
changed it to inputs

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+
+		List<ConversationInput> daprConversationInputs = new ArrayList<>();
+		daprConversationInputs.add(daprConversationInput);
+
```
Could we rename it to `response`

siri-varma replied:
renamed to response

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +546,142 @@ public void onError(RuntimeException exception) {
+		verify(daprStub, times(1)).converseAlpha1(captor.capture(), Mockito.any());
+
+		DaprProtos.ConversationRequest conversationRequest = captor.getValue();
+
```
Could we use static imports so we could use just `assertEquals` without `Assertions`

siri-varma replied:
Added static imports

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/DaprClientImpl.java:
```diff
@@ -1402,6 +1407,79 @@ public Mono<DaprMetadata> getMetadata() {
+            .setContent(input.getContent())
+            .setScrubPII(input.isScrubPii());
+
+        if (input.getRole() != null) {
```
`ConversationRole` is an enum with TOOL, USER, ASSISTANT and `toString()` would return capital case "TOOL", "USER", "ASSISTANT". While I think enum is a good idea, I would keep it compatible with GRPC object and use an ordinary string. We want users to decide what role they want to use `system`, `assistant`, `agent`, etc

siri-varma replied:
Changed it to string. If the tool, user and assistant are just examples https://docs.dapr.io/reference/api/conversation_api/#input-body I can update the readme here

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/ConversationRole.java:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ * Conversation AI supported roles.
+ */
```
As I mentioned I am not sure we need this enum. We want user to use whatever they feel is appropriate as a role. We don't want to keep changing this enum, because someone has a different use case.

cicoyle replied:
I replied on the other thread - I agree, lets align and use string. I appreciate the enum idea here though 🙏🏻

siri-varma replied:
I changed it to String from enum

https://docs.dapr.io/reference/api/conversation_api/#input-body
However in the Dapr Doc it states possible values are
```Possible values: ‘user’, ’tool’, ‘assistant’```

@cicoyle  Should this be updated examples of some values ?

cicoyle replied:
If we put string in the sdk that allows for us to support all types going forward with no additional work on our part, assuming more might get added upstream in the future.

I don't quite follow your question - could you restate it? I see the link you provided, and in the input-body we do have noted the potential values already. If you see a good spot to mention it elsewhere in the dapr/docs or in our sdk please feel free to add it!

artur-ciocanu reviewed (changes requested):
@siri-varma thanks a lot for all your hard work! I have a few general comments that are critical, others are mostly style related:
- Dapr Conversation API uses a component to configure everything related to model interaction and we should make sure that this is reflected in the API like using `name` or `componentName`. I feel `componentName` makes more sense, but I will let @cicoyle and @salaboy share their thoughts.
- `scrubPII` - I know this comes from Dapr Runtime, but honestly it looks and feels cryptic, so if we can agree on a better name and then propagate this back to Dapr Runtime that would be great.
- `ConversationRole` - any LLM framework I have used, do not constraint the role in any way, since you can have different roles, depending on use case, so I think sticking to Dapr Runtime string is good enough.

Please take a look and let me know your thoughts.

cicoyle commented:
> @siri-varma thanks a lot for all your hard work! I have a few general comments that are critical, others are mostly style related:
> 
> * Dapr Conversation API uses a component to configure everything related to model interaction and we should make sure that this is reflected in the API like using `name` or `componentName`. I feel `componentName` makes more sense, but I will let @cicoyle and @salaboy share their thoughts.
> * `scrubPII` - I know this comes from Dapr Runtime, but honestly it looks and feels cryptic, so if we can agree on a better name and then propagate this back to Dapr Runtime that would be great.
> * `ConversationRole` - any LLM framework I have used, do not constraint the role in any way, since you can have different roles, depending on use case, so I think sticking to Dapr Runtime string is good enough.
> 
> Please take a look and let me know your thoughts.

I understand where you are coming from, however we want to align with dapr runtime and other sdks. For example, the `scrubPII` should stay as-is, bc in [runtime](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto#L1322) we have it named as such, [dotnet](https://github.com/dapr/dotnet-sdk/blob/32d06a7136a4e32a70e02398619442d28f120165/src/Dapr.AI/Conversation/ConversationOptions.cs#L31) has it named that way, so for consistency in the dapr projects and ecosystem, the java sdk should as well. 

Same for component name - [dotnet](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto#L1307) && [go-sdk](https://github.com/dapr/go-sdk/blob/a3df75f17b648c0de4642fb67c575f3d34d2dc62/client/conversation.go#L26) for example

For `role` - Im inclined to agree on `string` as well for the same reasons. It is how we have it in the [go-sdk](https://github.com/dapr/go-sdk/blob/a3df75f17b648c0de4642fb67c575f3d34d2dc62/client/conversation.go#L50).

cicoyle commented on examples/src/main/java/io/dapr/examples/conversation/README.md:
```diff
@@ -0,0 +1,117 @@
+## Manage Dapr Jobs via the Conversation API
```
good catch, lets remove jobs from this unless you are using the `jobs api`.

siri-varma replied:
copy paste error. Fixed the error and thanks for catching it

cicoyle commented on pom.xml:
```diff
@@ -17,7 +17,7 @@
     <grpc.version>1.69.0</grpc.version>
     <protobuf.version>3.25.5</protobuf.version>
     <protocCommand>protoc</protocCommand>
```
`v1.15.4` is latest as of now - think this was just bc the PR has been open a while and we more recently released 1.15.3 && 1.15.4

siri-varma replied:
rebased with main which is using 1.15.4

siri-varma commented:
will update this pr in couple of days

siri-varma commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/TestDaprConversationConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+
+        return new DaprClientBuilder().withPropertyOverrides(overrides).buildPreviewClient();
+    }
+}
```
done

siri-varma commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/TestConversationApplication.java:
```diff
@@ -0,0 +1,26 @@
+        SpringApplication.run(TestConversationApplication.class, args);
+    }
+
+}
```
done

siri-varma commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprConversationIT.java:
```diff
@@ -0,0 +1,131 @@
+        Assertions.assertEquals("input this <PHONE_NUMBER>",
+                response.getConversationOutpus().get(1).getResult());
+    }
+}
```
done

siri-varma commented on examples/src/main/java/io/dapr/examples/conversation/DemoConversationAI.java:
```diff
@@ -0,0 +1,48 @@
+          + "This is the my number 672-123-4567");
+
+      // Component name is the name provided in the metadata block of the conversation.yaml file.
+      Mono<ConversationResponse> responseMono = client.converse(new ConversationRequest("echo",
```
done

siri-varma commented:
@dapr-reviewers  addressed all the comments

cicoyle commented:
> @dapr-reviewers addressed all the comments

I think you are looking for the following: @dapr/approvers-java-sdk @dapr/maintainers-java-sdk

If you type `java-` after the `@` it should show you the right ones :)

siri-varma commented:
@dapr/approvers-java-sdk rebased the branch and ready for review

siri-varma commented:
@artur-ciocanu , @cicoyle let me know if you folks have any more feedback on this. Thank you.

artur-ciocanu reviewed (dismissed):
@siri-varma thanks for your contribution, I have no more comments, great job!

artur-ciocanu commented:
@cicoyle and @salaboy could you please do a final pass. I have checked and everything looks good to me.

siri-varma commented:
@cicoyle I think this PR is ready to merge unless you or @salaboy have any additional feedback. Thank you

cicoyle commented on examples/src/main/java/io/dapr/examples/conversation/README.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+Get into the examples' directory:
````
last nit: can you remove this block to cd to the examples dir since its already noted above?

siri-varma replied:
https://github.com/dapr/java-sdk/pull/1235/commits/f994c8a6add96cdd8aee24db1a52724d4938498f done removed the block
